cmd/ci-operator/main.go

```diff
@@ -1212,3 +1210,65 @@ func eventRecorder(kubeClient *coreclientset.CoreV1Client, namespace string) rec
 	return eventBroadcaster.NewRecorder(
 		templatescheme.Scheme, coreapi.EventSource{Component: namespace})
 }
+
+func getSSHSecretFromPath(sshKeyPath string) (*coreapi.Secret, error) {
+	secret := &coreapi.Secret{Data: make(map[string][]byte)}
+	secret.Type = coreapi.SecretTypeSSHAuth
+	secret.Name = convertToValidSecretName(filepath.Base(sshKeyPath))
+
+	files, err := ioutil.ReadDir(sshKeyPath)
+	if err != nil {
+		return nil, fmt.Errorf("could not read dir %s for secret: %v", sshKeyPath, err)
+	}
+
+	for _, f := range files {
+		if f.IsDir() {
+			continue
+		}
+		path := filepath.Join(sshKeyPath, f.Name())
+		// if the file is a broken symlink or a symlink to a dir, skip it
+		if fi, err := os.Stat(path); err != nil || fi.IsDir() {
+			continue
+		}
+
+		// Secret.Data["ssh-privatekey"] is required on SecretTypeSSHAuth type.
+		// https://github.com/kubernetes/api/blob/master/core/v1/types.go#L5466-L5470
+		secret.Data[coreapi.SSHAuthPrivateKey], err = ioutil.ReadFile(path)
+		if err != nil {
+			return nil, fmt.Errorf("could not read file %s for secret: %v", path, err)
+		}
+	}
+	return secret, nil
+}
+
+func createORUpdateSecret(namespace string, client *coreclientset.CoreV1Client, secret *coreapi.Secret) error {
+	_, err := client.Secrets(namespace).Create(secret)
+	if kerrors.IsAlreadyExists(err) {
+		existing, err := client.Secrets(namespace).Get(secret.Name, meta.GetOptions{})
+		if err != nil {
+			return fmt.Errorf("could not retrieve secret %s: %v", secret.Name, err)
+		}
+		if l := len(secret.Data); l != 0 && existing.Data == nil {
+			existing.Data = make(map[string][]byte, l)
+		}
+		for k, v := range secret.Data {
+			existing.Data[k] = v
+		}
+		if _, err := client.Secrets(namespace).Update(existing); err != nil {
+			return fmt.Errorf("could not update secret %s: %v", secret.Name, err)
+		}
+		log.Printf("Updated secret %s", secret.Name)
+		return nil
+	}
+
+	if err != nil {
+		return fmt.Errorf("could not create secret %s: %v", secret.Name, err)
+	}
+	log.Printf("Created secret %s", secret.Name)
+	return nil
+}
+
+func convertToValidSecretName(secretName string) string {
+	re := regexp.MustCompile(`[^a-z0-9]([^-a-z0-9]*[^a-z0-9])?(\.[^a-z0-9]([^-a-z0-9]*[^a-z0-9])?)*`)
```

Comment:
We should use upstream validators and mutators for this

Reply:
Or just reject invalid names at runtime and not try to change anything here

Reply:
Yea, my point here is to convert any not-allowed char to `-`. However, I will try to check if there is an upstream validator.